Stash the subscriber list when we disconenct from the bus

rpigott opened the pull request:
If we unexpectly disconnect from the bus, systemd would end up dropping
the list of subscribers, which breaks the ability of clients like logind
to monitor the state of units.

Stash the list of subscribers into the deserialized state in the event
of a disconnect so that when we recover we can renew the broken
subscriptions.

---

Fixes: #8672 #26744

msekletar reviewed (commented):
Isn't there a problem that `m->deserialized_subscribed` is single list and we put into it bus tracking information from all buses tracked by `Manager`, i.e. `destroy_bus()` is called for system, API and private buses. However, we later use this stash/list only on api bus, because `bus_track_coldplug()` is called only for that bus.

rpigott commented:
I don't think so. `m->subscribed` only tracks the api bus, and `destroy_bus` only stashes from `m->subscribed`:

https://github.com/systemd/systemd/blob/144d291f1809cc9070eae89155db9f1a4ddb8625/src/core/dbus.c#L1001

This matches the usage in `manager_serialize`.

keszybz commented on src/core/dbus.c:
```diff
@@ -131,10 +131,15 @@ static int signal_disconnected(sd_bus_message *message, void *userdata, sd_bus_e
+        }
+        if (bus == m->system_bus) {
                 bus_done_system(m);
+                if (!MANAGER_IS_SYSTEM(m))
+                        log_debug("Got disconnect on system bus.");
```
Hmm, above the logging is before `bus_done_api()`, and here it is after `bus_done_system()`. I think it should be before.

Why is the logging excluded when `MANAGER_IS_SYSTEM()`? Maybe add an explanatory comment.

rpigott replied:
If `MANAGER_IS_SYSTEM`, the api bus and the system bus are the same, so we don't need to log about it twice.

keszybz commented on src/core/dbus.c:
```diff
@@ -984,6 +988,19 @@ int bus_init_private(Manager *m) {
+        for (const char *n = sd_bus_track_first(t); n; n = sd_bus_track_next(t)) {
+                int c, j;
+
+                c = sd_bus_track_count_name(t, n);
+                for (j = 0; j < c; j++)
+                        if (strv_extend(l, n) < 0)
+                                log_oom();
```
Hmm, so we use repeated strings in the strv to specify the count. This seems iffy. Why not use a hashmap name→count instead?

Edit: Oh, OK, this is because that's how `deserialized_subscribed` is used already. I guess that's OK, if the repeat counts are not too high.

keszybz commented on src/core/dbus.c:
```diff
@@ -984,6 +988,19 @@ int bus_init_private(Manager *m) {
+                int c, j;
+
+                c = sd_bus_track_count_name(t, n);
+                for (j = 0; j < c; j++)
```
Please move `int j` into the loop header.

keszybz commented on src/core/manager.c:
```diff
@@ -1803,6 +1803,9 @@ Manager* manager_free(Manager *m) {
         free(m->switch_root_init);
 
+        sd_bus_track_unref(m->subscribed);
+        strv_free(m->deserialized_subscribed);
```
I think it'd prefer to rename this field. (`inherited_subscribed` or `previously_subscribed` or something.)  I think it'll be confusing if it's called "deserialized", but actually it's set at runtime.

rpigott replied:
Well, it _is_ serialized. This is just the already existing field from before additionally used in the "bus disconnect" case — I think I'd prefer not to change the name.

I added these two lines because these fields actually were just not previously freed in manager free, which seems like an accidental omission to me either way. However, this did not previously cause any leak afaict because no code path reached manager free before freeing these fields. With these changes, it potentially could, so they must now be freed here for correctness.

poettering replied:
i agree with @keszybz please rename this, this is too confusing. maybe renamed it "subscribe_serialization" or so?

keszybz reviewed (commented):
Looks reasonable. How do you test this?

rpigott commented:
> Looks reasonable. How do you test this?

I first reproduced the issue with the reproducer in #32381, but it's kind of random, so when testing this patch I hacked systemd to respond to SIGUSR1 by killing it's own bus connection, similar to the behavior when a timeout occurs in sd_bus_add_match_async. 

I don't think you could observe the subscriber list anywhere previously, so I added a commit to dump it in the analyze-dump output. I watched this output and added pam_systemd.so to my sudo pam config temporarily so that systemd-run sudo whoami would create a session, which would then get leaked into a permanent closing state as reported in #8672.

rpigott commented:
I bumped the level of log messages about disconnect. I don't think this ever actually an expected condition, even if we can successfully recover, and _probably_ indicates something else is going wrong. So s/log_debug/log_notice/ for those.

poettering commented:
hmm, if we drop from the bus (which is a case that *really* shouldn't happen btw, it's a failure scenario that is going to break so much shit, and is probably as "unthinkable" as dbus daemon being restarted) then there's going to be time window where we are not going to keep the list updated, i.e. we do not drop entries from it that are invalidated (because a peer also dropped off the bus – which is quite likely as many dbus clients are shortlived and if the system is in such a bad state that pid 1 got kicked off the bus, then most likely other stuff also goes down and gets kicked off). but that means the peer list is going to be stuffed with out-of-date stuff, which we really need to revalidate.

i wonder if the strategy shouldn't be the inverse here: make the clients re-subscribe when we come back rather than keep the subscriber list in pid1. That would make the issue go away: we always start from a clean state.

but of course this would move the problem to clients, i.e. out of a single place into many places, which isn't that great either.

poettering commented on src/core/dbus.c:
```diff
@@ -854,6 +860,10 @@ int bus_init_api(Manager *m) {
         if (r < 0)
                 return log_error_errno(r, "Failed to set up API bus: %m");
 
+        r = bus_track_coldplug(bus, &m->subscribed, false, m->deserialized_subscribed);
```
our current coding style suggests to put parameter comments before opaque parameters, i.e. the third param here should be prefied with `/* recursive= */`

poettering commented on src/core/dbus.c:
```diff
@@ -979,6 +989,19 @@ int bus_init_private(Manager *m) {
+                c = sd_bus_track_count_name(t, n);
+                for (int j = 0; j < c; j++)
+                        if (strv_extend(l, n) < 0)
+                                log_oom();
```
propagate the error?

poettering commented on src/core/dbus.c:
```diff
@@ -979,6 +989,19 @@ int bus_init_private(Manager *m) {
+        for (const char *n = sd_bus_track_first(t); n; n = sd_bus_track_next(t)) {
+                int c;
+
+                c = sd_bus_track_count_name(t, n);
```
this can fail

poettering commented on src/core/dbus.c:
```diff
@@ -979,6 +989,19 @@ int bus_init_private(Manager *m) {
+        for (const char *n = sd_bus_track_first(t); n; n = sd_bus_track_next(t)) {
+                int c;
+
+                c = sd_bus_track_count_name(t, n);
```
hmm, so subscription is not recursive, we always pass "false" in the "recurse" parameter when allocating the bus track object.

hence sd_bus_track_count_name() should return 1 always here, hence you can smplify this...

rpigott replied:
I'll do the generic bus_track_to_strv thing then I guess.

poettering commented on src/core/dbus.c:
```diff
@@ -979,6 +989,19 @@ int bus_init_private(Manager *m) {
+        }
+
+        return 0;
+}
```
or alternatively, this seems somewhat generic, hence maybe move this into src/shared/bus-util.[ch] and then do keep the sd_bus_track_count_name() stuff in place. call it bus_track_to_strv() or so?

poettering commented on src/core/dbus.c:
```diff
@@ -979,6 +989,19 @@ int bus_init_private(Manager *m) {
+
+                c = sd_bus_track_count_name(t, n);
+                for (int j = 0; j < c; j++)
+                        if (strv_extend(l, n) < 0)
```
in systemd we prefer to have output params that are only initialized on success. i.e. please build the strv locally first, and then on access move them into a `ret` parameter.

rpigott commented:
> hmm, if we drop from the bus (which is a case that really shouldn't happen btw, it's a failure scenario that is going to break so much shit, and is probably as "unthinkable" as dbus daemon being restarted)

If dropping the bus connection is really "unthinkable" then the sd_bus_add_match_async default install callback, which just kills the bus connection, really doesn't seem acceptable. It should probably be removed from the codebase everywhere but it is used not infrequently afaict. E.g. See also #35450.

rpigott commented:
> i wonder if the strategy shouldn't be the inverse here: make the clients re-subscribe when we come back rather than keep the subscriber list in pid1. That would make the issue go away: we always start from a clean state.

I considered that, but isn't there a compat issue? Subscribe doesn't currently have anyway to tell clients they need to resubscribe and tbh I don't know how it would be implemented.

rpigott commented:
I don't think 4 makes sense without 3? I amended the function of the deserialized_subscribed member. If it is unchanged, there's no reason to modify the name. I can split just 1&2 if you'd like I guess though, but #8672 should get re-opened if we don't address it. I have demonstrated it is reproducible on the current master.

Anyway what is the objection to 3? I understand that there will be a brief window where we aren't connected to the bus, but isn't it _much_ better than the status quo? To address @poettering's concern:

> we do not drop entries from it that are invalidated (because a peer also dropped off the bus – which is quite likely as many dbus clients are shortlived and if the system is in such a bad state that pid 1 got kicked off the bus, then most likely other stuff also goes down and gets kicked off). but that means the peer list is going to be stuffed with out-of-date stuff, which we really need to revalidate.

I don't think that's true? Doesn't sd_bus_track_add_name check that the unique name actually exists when it subscribes?

https://github.com/systemd/systemd/blob/d218ecc7337a4689e7d589df7fd146f4aaab3809/src/libsystemd/sd-bus/bus-track.c#L242-L244

Or do you expect that a unique name will get reused by the dbus daemon during the period we are disconnected?

The consequences of leaking sessions is pretty severe imo. Yes, we should take steps to stop systemd from closing the bus connection, but we might as well do our best to recover when we fail.

bluca commented:
Just trying to get things moving, I have no opinion one way or the other. First 2 commits then?

rpigott commented:
Ok then, #35603 has the first 2 commits only.

rpigott commented:
@poettering ping?

This still seems to happen in the wild https://bbs.archlinux.org/viewtopic.php?id=301983.

poettering commented on src/core/dbus.c:
```diff
@@ -860,6 +860,10 @@ int bus_init_api(Manager *m) {
         if (r < 0)
                 return log_error_errno(r, "Failed to set up API bus: %m");
 
+        r = bus_track_coldplug(bus, &m->subscribed, /* recursive= */ false, m->deserialized_subscribed);
```
bus_track_coldplug will fail if any of the provided names disappeared in the mean time. i think we should just ignore failures here, i.e. cast the result to (void)

poettering commented on src/core/dbus.c:
```diff
@@ -1004,8 +1008,19 @@ static void destroy_bus(Manager *m, sd_bus **bus) {
+
+                r = bus_track_to_strv(m->subscribed, &subscribed);
+                if (r < 0)
+                        log_error_errno(r, "Failed to serialize api subscribers: %m");
```
i think it is fine to ignore failures here, but in that case downgrade this to `log_warning`, and say in the log message that you are ignoring this failure, i.e. end it in `… subscribers, ignoring: %m");`

poettering commented on src/core/dbus.c:
```diff
@@ -1004,8 +1008,19 @@ static void destroy_bus(Manager *m, sd_bus **bus) {
+                        log_error_errno(r, "Failed to serialize api subscribers: %m");
+                r = strv_extend_strv_consume(&m->deserialized_subscribed, TAKE_PTR(subscribed), /* filter_duplicates = */ false);
+                if (r < 0)
+                        log_error_errno(r, "Failed to record api subscribers: %m");
```
similar here

poettering commented on src/core/dbus.c:
```diff
@@ -860,6 +860,10 @@ int bus_init_api(Manager *m) {
+        r = bus_track_coldplug(bus, &m->subscribed, /* recursive= */ false, m->deserialized_subscribed);
+        if (r < 0)
+                return r;
+        m->deserialized_subscribed = strv_free(m->deserialized_subscribed);
```
i'd really rename the field, now that we not just using it during deserialization, but also during serialization. i.e. maybe rename the field to `subscribed_as_strv` or so?

poettering replied:
ah, the 2nd patch is doing that.

rpigott replied:
I did change it to "subscribed_serialization" as suggested by https://github.com/systemd/systemd/pull/35406#discussion_r1880525310, but in a following commit to separate the logical change from the cosmetic one. I can squash them if you'd like. `subscribed_strv` sounds better to me now though.

rpigott replied:
I made it `subscribed_as_srtv` now.

poettering commented on src/core/dbus.c:
```diff
@@ -1004,8 +1008,19 @@ static void destroy_bus(Manager *m, sd_bus **bus) {
+                r = bus_track_to_strv(m->subscribed, &subscribed);
+                if (r < 0)
+                        log_error_errno(r, "Failed to serialize api subscribers: %m");
+                r = strv_extend_strv_consume(&m->deserialized_subscribed, TAKE_PTR(subscribed), /* filter_duplicates = */ false);
```
why append here btw? shouldn't we just replace? can it ever happen we actually merge two populated lists here?

rpigott replied:
I'm not sure. I guess I thought this would handle either case OK.

poettering replied:
i would prefer if we'd not extend here, but replace. for the simple reason we should not have something that by accident can grow without boundaries. If there's no reason to extend here, then please just replace.

poettering commented:
so the sd-bus-track logic already validates each name you pass to it, hence I think my revalidation concerns are at least 50% dealt with. I think this is good enough to go in, but see some comments above.

The other 50% of my concerns go something like this: there are two cases where we can be disconnected from the bus.

1. we are disconnected, but the bus stays running. this case is covered well enough with the logic above, as we track clients by unique name, and those should be unique as long as the bus is around
2. the bus goes down and we are disconnected as a result. this case is not covered at all by this. in this case the unique names will start from zero, and we thus populate the bus tracker with a bunch of unique names that might reference completely different clients as before.

i guess the 2nd thing could be addressed by the bus daemon instead of us. for example they could serialize their unique id counter or so. or derive it from the invocation id or so...

rpigott commented:
Wait, is case #2 recoverable at all? I thought if dbus went down the system was basically cooked.

poettering commented:
> Wait, is case #2 recoverable at all? I thought if dbus went down the system was basically cooked.

well, in the big picture apps won't like it if the bus disappears, hence yes, it's illusionary having a full blown desktop survive if the system bus goes away. that said, I am not opposed to making systemd itself work reasonably well if this happens.

poettering commented:
i filed this btw:

https://github.com/bus1/dbus-broker/issues/383

poettering commented:
ok, if you make this replace rather than extend the strv, seems ok to go in. consider using strv_free_and_replace()

poettering commented:
as mentioned a follow-up PR that queries the bus id from the bus driver before reinstalling the tracking would be very welcome. hint hint ;-)

rpigott commented:
Yeps sounds like a good idea. I don't mind implementing that in a future PR but I'm sick and don't really feel like it rn :sweat:.

YHNdnzj commented on src/shared/bus-util.c:
```diff
@@ -698,6 +698,28 @@ int bus_track_add_name_many(sd_bus_track *t, char **l) {
+        assert(ret);
+
+        for (const char *n = sd_bus_track_first(t); n; n = sd_bus_track_next(t)) {
+                r = sd_bus_track_count_name(t, n);
+                if (r < 0)
+                        return r;
+
+                for (int j = 0; j < r; j++) {
+                        r = strv_extend(&subscribed, n);
+                        if (r < 0)
+                                return r;
+                }
```
Hmm? Seems spurious that you store the count in `r` but immediately override it with the result of strv_extend?

YHNdnzj commented:
I'll submit a follow-up addressing the post-review comment and the bus id thing

rpigott commented:
thanks
